# Geocodificar usuarios [Base completa]

In [1]:
import pandas as pd
import get_centroides
from get_centroides import getFromDict
import sqlalchemy
import warnings
import pickle

In [2]:
df = pd.read_csv('./HI_usuarios_completos.csv',  encoding='latin-1') # Datos institucion
df.head()

anio_ing nacionalidad provincia_part    loc_part cp_part
0      2014           AR             CF        CABA    1416
1      2014           AR             CF        CABA    1416
2      2014           AR             BA  HURLINGHAM    1686
3      2014           AR             BA      ZARATE    2800
4      2014           AR             CF        CABA    1425

In [3]:
df.dtypes

anio_ing           int64
nacionalidad      object
provincia_part    object
loc_part          object
cp_part           object
dtype: object

In [4]:
df.nacionalidad.unique() # Algunos NaN y fechas en el campo nacionalidad.

array(['AR', 'CA', 'UY', 'PY', 'IT', 'US', 'AU', 'BR', 'JP', 'BO', 'CN',
       'ES', 'CO', 'VE', 'KR', 'EC', 'PE', 'CL', 'FR', 'GB', 'HL', 'PT',
       'CU', 'SY', 'KP', 'PA', 'CR', 'YU', 'HN', 'PL', 'UA', 'DE', 'MX',
       'GT', 'IL', 'HT', 'AO', 'RU', '30/06/2015', 'IO', 'CH', 'HR', 'AT',
       'LT', 'HU', 'TH', 'ET', '05/08/1987', 'DO', 'IN', 'TW', 'MK', 'SV',
       'NZ', 'GR', 'AM', 'DM', 'SK', 'PH', 'SE', 'BG', 'VN', 'DK', 'SI',
       'NI', 'IR', 'FI', 'IE', 'KZ', 'TR', 'LI', nan, 'BE', 'LB', 'BY',
       'IS', 'ZA', 'NG', 'MZ', 'BD', 'PS'], dtype=object)

In [5]:
df.provincia_part.unique() # Parece ser que provincia es un campo estructurado

array(['CF', 'BA', nan, 'LP', 'MZ', 'SL', 'CA', 'SC', 'CH', 'RN', 'TF',
       'SF', 'ER', 'MI', 'CB', 'TU', 'NE', 'CO', 'JU', 'SJ', 'SA', 'FO',
       'SE', 'CT'], dtype=object)

In [6]:
df.nacionalidad.str.contains('/').sum() # Tengo el mismo problema que antes

2

In [7]:
df.isna().sum() 
# 206 usuarios sin  nacionalidad
# 4820 sin provincia reportada
# 1 solo sin CP    

anio_ing             0
nacionalidad       206
provincia_part    4820
loc_part           263
cp_part              1
dtype: int64

Pareceria que todo es muy similar al dataset mas pequeño con el que trabajamos antes. Exploraremos algunas aspectos mas granulares (ej por nacionalidad) parar ver si en los casos mas 'anomalos' hay algun conflicto.

In [8]:
exp_gb = df.groupby(['anio_ing','nacionalidad','provincia_part']).agg('count').reset_index()

In [9]:
exp_gb.isna().sum() # Los datos de gente de otra nacionalidad parecerian estar todos completos

anio_ing          0
nacionalidad      0
provincia_part    0
loc_part          0
cp_part           0
dtype: int64

In [10]:
exp_gb[exp_gb.nacionalidad != 'AR'].provincia_part.unique() 
# Parece que todos las personas que reportan otra nacionalidad igualmente seleccionan una provincia
# en el territorio argentino

array(['BA', 'CF', 'SA', 'RN', 'CO', 'MI', 'CA'], dtype=object)

## Preprocessing

Vamos a remover los dos sujetos que presentan fechas en las nacionalidades. 

In [11]:
# Vamos a asumir que los campos con valor None en nacionalidad se deben a errores de los usuarios, pasaran a ser `desconocido`.
# Los campos que tengan fechas en la nacionalidad seran removidos.
df.nacionalidad[df.nacionalidad.isna()] = 'DESCONOCIDO'

C:\Users\joaqu\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [12]:
df_clean = df.loc[df.nacionalidad.str.contains('/') == False].reset_index(drop = True).copy()

In [13]:
df_clean.shape[0] - df.shape[0] # Eliminamos dos usuarios de 80k

-2

Para este caso asignare un `user_id` simulando un dataset real. 

In [14]:
df_clean['user_id'] = list(map(lambda x: 'ID_' + str(x), df_clean.index.to_list()))

In [15]:
# Poner loc-part todo en mayuscula
df_clean.loc_part = df_clean.loc_part.apply(lambda x: x.upper() if isinstance(x,str) else None)  

## Acceder a la base de datos 

In [16]:
warnings.simplefilter("ignore", category=sqlalchemy.exc.SAWarning) # ignor sqlalchemy warnings
engine = sqlalchemy.create_engine('sqlite:///SQL/CPA.db')
con = engine.connect()
cpa_prov = pd.read_sql('provincia', con=con) # provincia por codigo postal
cpa_local = pd.read_sql('localidad', con=con) # localidad por codigo postal
#Se debe remover este campo del dataframe antes de realizar el join.
cpa_local.drop(cpa_local[cpa_local.codigopostal == 0].index).reset_index(drop = True, inplace=True)
cpa_capital =  pd.read_sql('comunas_capital', con=con) # localidades y comuna de capital por CP

In [17]:
cpa_provincia_completo = cpa_local.merge(cpa_prov, left_on='provincia_id', right_on='p_id',
                                         suffixes=['_localidad','_provincia']).drop(columns = 'p_id')
CPA_ARGENTINA =  cpa_provincia_completo.merge(cpa_capital, on= 'provincia_id' ,
                                              how='left' , suffixes = ('_x',''))  
# El dataframe mas completo hasta el momento de CPA y localidad
# Unir los dos codigos postales de las distintas tablas
CPA_ARGENTINA['codigopostal'] = CPA_ARGENTINA.codigopostal.fillna(CPA_ARGENTINA.codigopostal_x)
CPA_ARGENTINA.drop(columns='codigopostal_x', inplace= True) # eliminar la columna CP repetida
CPA_ARGENTINA['codigopostal'] = CPA_ARGENTINA.codigopostal.astype(int) # Transformar a integer el CP

In [18]:
set_provincia = set(cpa_provincia_completo.codigopostal)
set_provincia.remove(0) # Nuevamente removemos el 0 del codigo postal de provincia
set_capital = set(cpa_capital.codigopostal)
set_capital.intersection(set_provincia) # No hay CP en capital que se repitan en las otras provincias

set()

In [19]:
## Acceder a los datos de la API
departamentos = pd.read_csv('https://apis.datos.gob.ar/georef/api/departamentos.csv')
localidades = pd.read_csv('https://apis.datos.gob.ar/georef/api/localidades.csv')
municipios = pd.read_csv('https://apis.datos.gob.ar/georef/api/municipios.csv')
provincias = pd.read_csv('https://apis.datos.gob.ar/georef/api/provincias.csv')

## Populando nueva tabla con datos de geolocalizacion
### Asignar localidad, departamento(comuna CP) y provincia usando  solo CP

In [20]:
# Listado de CP que no son unicos
provincia_por_cp = cpa_provincia_completo.groupby('codigopostal')['nombre_provincia'].nunique()
cp_prov_no_unicos = set(provincia_por_cp.loc[provincia_por_cp > 1].index)

localidad_por_cp = cpa_provincia_completo.groupby('codigopostal')['nombre_localidad'].nunique()
cp_localidad_no_unicos = set(localidad_por_cp.loc[localidad_por_cp > 1].index) 

capital_barrio_por_cp = cpa_capital.groupby('codigopostal')['barrio'].nunique()
capital_cp_barrio_no_unicos = set(capital_barrio_por_cp.loc[capital_barrio_por_cp > 1].index)

capital_comuna_por_cp = cpa_capital.groupby('codigopostal')['comuna'].nunique()
capital_cp_comuna_no_unicos = set(capital_comuna_por_cp.loc[capital_comuna_por_cp > 1].index)

In [21]:
# Funciones utiles
def infer_provincia(cp):
    try:
        int(cp)
    except:
        return 'CP INDEFINIDO'
    if int(cp) in set_capital:
        return 'CIUDAD AUTONOMA DE BUENOS AIRES'
    if int(cp) in set_provincia:
        if  int(cp) in cp_prov_no_unicos:
            return  'NO UNICA PROVINCIA'
        else:
            prov = cpa_provincia_completo.loc[cpa_provincia_completo.codigopostal == int(cp)].nombre_provincia.unique()
            return prov[0].upper()
    if (int(cp) not in set_provincia and int(cp) not in set_capital):
        return 'CP INDEFINIDO'
    if cp.isna() or int(cp) == 0:
        return 'CP INDEFINIDO'

def infer_localidad(cp):
    try:
        int(cp)
    except:
        return 'CP INDEFINIDO'
    cp = int(cp)
    if cp in set_capital:
        if cp in capital_cp_barrio_no_unicos:
                return 'NO UNICA LOCALIDAD'
        else:
            localidad = cpa_capital.loc[cpa_capital.codigopostal == cp].barrio.unique()[0]
            return localidad.upper()
    if cp in set_provincia:
        if cp in cp_localidad_no_unicos:
            return 'NO UNICA LOCALIDAD'
        else:
            localidad = cpa_provincia_completo.loc[cpa_provincia_completo.codigopostal == cp].nombre_localidad.unique()[0]
            return localidad.upper()
    if cp not in set_provincia and cp not in set_capital:
        return 'CP INDEFINIDO'
           
def infer_departamento(cp):
    try:
        int(cp)
    except:
        return 'CP INDEFINIDO'
    cp = int(cp)
    if cp in set_capital:
        if cp in capital_cp_barrio_no_unicos:
            return 'NO UNICO DEPARTAMENTO'
        else:
            departamento = cpa_capital.loc[cpa_capital.codigopostal == cp].comuna.unique()
            return f'COMUNA {departamento[0]}'

In [22]:
augmented_df = df_clean.copy() # este es el dataframe que vamos a popular
#Agregamos columnas
columnas =  ['provincia_por_cp' ,'localidad_por_cp' ,'departamento_por_cp' ,'localidad_por_API', 'localidad_id_API', 'municipio_por_API',
             'municipio_id_API', 'departamento_por_API', 'departamento_id_API', 'provincia_por_API', 'provincia_id_API','centroide_lat_localidad',
             'centroide_lon_localidad', 'centroide_lat_municipio','centroide_lon_municipio','centroide_lat_departamento','centroide_lon_departamento',
             'centroide_lat_provincia','centroide_lon_provincia']
for i in columnas:
    augmented_df[i] = None

In [23]:
# Provincia
augmented_df['provincia_por_cp'] = augmented_df.cp_part.apply(infer_provincia)
# Localidad
augmented_df['localidad_por_cp'] = augmented_df.cp_part.apply(infer_localidad)
# Departamento 'Comuna'
augmented_df['departamento_por_cp'] = augmented_df.cp_part.apply(infer_departamento)

In [24]:
 augmented_df.sample(10)

anio_ing nacionalidad provincia_part             loc_part cp_part  \
1010       2014           AR             BA    JOSE CLEMENTE PAZ    1665   
37724      2016           AR             BA      VALENTIN ALSINA    1822   
27937      2016           AR             BA               OLIVOS    1636   
62741      2018           AR             CF                 CABA    1429   
77786      2019           AR             CF  CDAD AUTONOMA BS AS    1200   
52951      2017           AR             BA               OLIVOS    1602   
11995      2014           AR             BA      VILLA BALLESTER    1653   
48679      2017           AR             BA           SAN ANDRES    1651   
83468      2019           AR             BA  GREGORIO DE LA FERR    1757   
66626      2018           AR             BA              CASEROS    1678   

        user_id                 provincia_por_cp       localidad_por_cp  \
1010    ID_1010                     BUENOS AIRES      JOSE CLEMENTE PAZ   
37724  ID_37724                     BUENOS AIRES        VALENTIN ALSINA   
27937  ID_27937                     BUENOS AIRES     NO UNICA LOCALIDAD   
62741  ID_62741  CIUDAD AUTONOMA DE BUENOS AIRES     NO UNICA LOCALIDAD   
77786  ID_77786                    CP INDEFINIDO          CP INDEFINIDO   
52951  ID_52951                     BUENOS AIRES     NO UNICA LOCALIDAD   
11995  ID_11995                     BUENOS AIRES        VILLA BALLESTER   
48679  ID_48679                     BUENOS AIRES             SAN ANDRES   
83468  ID_83468                     BUENOS AIRES  GREGORIO DE LAFERRERE   
66626  ID_66626                     BUENOS AIRES                CASEROS   

         departamento_por_cp localidad_por_API  ... provincia_por_API  \
1010                    None              None  ...              None   
37724                   None              None  ...              None   
27937                   None              None  ...              None   
62741  NO UNICO DEPARTAMENTO              None  ...              None   
77786                   None              None  ...              None   
52951                   None              None  ...              None   
11995                   None              None  ...              None   
48679                   None              None  ...              None   
83468                   None              None  ...              None   
66626                   None              None  ...              None   

      provincia_id_API centroide_lat_localidad centroide_lon_localidad  \
1010              None                    None                    None   
37724             None                    None                    None   
27937             None                    None                    None   
62741             None                    None                    None   
77786             None                    None                    None   
52951             None                    None                    None   
11995             None                    None                    None   
48679             None                    None                    None   
83468             None                    None                    None   
66626             None                    None                    None   

      centroide_lat_municipio centroide_lon_municipio  \
1010                     None                    None   
37724                    None                    None   
27937                    None                    None   
62741                    None                    None   
77786                    None                    None   
52951                    None                    None   
11995                    None                    None   
48679                    None                    None   
83468                    None                    None   
66626                    None                    None   

      centroide_lat_departamento centroide_lon_departamento  \
1010                        None  

### Conseguir datos de los que tienen provincia y localidad provista por codigo postal

In [25]:
# Seleccionamos solo las prov con loc
gby_prov_loc = augmented_df.groupby(['provincia_por_cp','localidad_por_cp']).count().reset_index(level=['provincia_por_cp', 'localidad_por_cp']).copy()
gby_prov_loc = gby_prov_loc.loc[(gby_prov_loc.localidad_por_cp != 'NO UNICA LOCALIDAD') & (gby_prov_loc.localidad_por_cp != 'CP INDEFINIDO')]

In [26]:
gby_prov_loc.drop(columns = ['anio_ing','user_id','nacionalidad','provincia_part','cp_part','loc_part','departamento_por_cp'], inplace= True)
gby_prov_loc.loc[:,'localidad_por_API' : 'centroide_lon_provincia'] = None # inicializo todos los campos a zero
print(gby_prov_loc.shape)
gby_prov_loc.head(4) # Tenemos 105 combinacinos unicas no identificadas

(105, 18)


provincia_por_cp   localidad_por_cp localidad_por_API localidad_id_API  \
0     BUENOS AIRES         9 DE ABRIL              None             None   
1     BUENOS AIRES  AEROPUERTO EZEIZA              None             None   
2     BUENOS AIRES     ALEJANDRO KORN              None             None   
3     BUENOS AIRES           BANFIELD              None             None   

  municipio_por_API municipio_id_API departamento_por_API departamento_id_API  \
0              None             None                 None                None   
1              None             None                 None                None   
2              None             None                 None                None   
3              None             None                 None                None   

  provincia_por_API provincia_id_API centroide_lat_localidad  \
0              None             None                    None   
1              None             None                    None   
2              None             None                    None   
3              None             None                    None   

  centroide_lon_localidad centroide_lat_municipio centroide_lon_municipio  \
0                    None                    None                    None   
1                    None                    None                    None   
2                    None                    None                    None   
3                    None                    None                    None   

  centroide_lat_departamento centroide_lon_departamento  \
0                       None                       None   
1                       None                       None   
2                       None                       None   
3                       None                       None   

  centroide_lat_provincia centroide_lon_provincia  
0                    None                    None  
1                    None                    None  
2                    None                    None  
3                    None                    None

In [27]:
# Crea una lista de tuplas para realizar el post. Para mas info consultar get_centroides.py
prov_loc = list(gby_prov_loc.apply(lambda x: (x.provincia_por_cp, x.localidad_por_cp), axis=1))
# Manda el POST request
datos = get_centroides.post_centroides('localidades', prov_loc, prov = True) # Consulta get_centroides.py

In [28]:
#funcion para rellenar los df
def rellenar_df(df,datos):
    df['provincia_por_API'] = [getFromDict(i,'provincia','nombre') for i in datos]
    df['provincia_id_API'] = [getFromDict(i,'provincia','id') for i in datos]
    df['departamento_por_API'] = [getFromDict(i,'departamento','nombre') for i in datos]
    df['departamento_id_API'] = [getFromDict(i,'departamento','id') for i in datos]
    df['municipio_por_API'] = [getFromDict(i,'municipio','nombre') for i in datos]
    df['municipio_id_API'] = [getFromDict(i,'municipio','id') for i in datos]
    df['localidad_por_API'] = [getFromDict(i,'nombre') for i in datos]
    df['localidad_id_API'] = [getFromDict(i,'id') for i in datos]
    df['centroide_lat_localidad'] = [getFromDict(i,'centroide','lat') for i in datos]
    df['centroide_lon_localidad'] = [getFromDict(i,'centroide','lon') for i in datos]
    cols_s = ['municipio_por_API', 'departamento_por_API', 'provincia_por_API']
    cols_i = ['departamento_id_API','municipio_id_API','provincia_id_API']
    # To upper
    for i in cols_s:
        df[i] = df[i].apply(lambda x: x.upper() if isinstance(x,str) else 'NO HUBO REQUEST')
    for i in cols_i:
        df[i] = df[i].apply(lambda x: (int(x) if  (x != 'NO HUBO REQUEST') and pd.notna(x) else x)) 
    return df

In [29]:
gby_prov_loc = rellenar_df(gby_prov_loc,datos)
gby_prov_loc.sample(5)

provincia_por_cp      localidad_por_cp  \
104  CIUDAD AUTONOMA DE BUENOS AIRES           SAN NICOLÁS   
0                       BUENOS AIRES            9 DE ABRIL   
67                      BUENOS AIRES  SIERRA DE LA VENTANA   
29                      BUENOS AIRES    ISLA MARTIN GARCIA   
40                      BUENOS AIRES            LONGCHAMPS   

        localidad_por_API localidad_id_API   municipio_por_API  \
104           SAN NICOLAS      02007010005     NO HUBO REQUEST   
0              9 DE ABRIL      06260010005  ESTEBAN ECHEVERRÍA   
67   SIERRA DE LA VENTANA      06819030000           TORNQUIST   
29        NO HUBO REQUEST  NO HUBO REQUEST     NO HUBO REQUEST   
40             LONGCHAMPS      06028010007     ALMIRANTE BROWN   

    municipio_id_API departamento_por_API departamento_id_API  \
104             None             COMUNA 1                2007   
0              60260   ESTEBAN ECHEVERRÍA                6260   
67             60819            TORNQUIST                6819   
29   NO HUBO REQUEST      NO HUBO REQUEST     NO HUBO REQUEST   
40             60028      ALMIRANTE BROWN                6028   

                   provincia_por_API provincia_id_API centroide_lat_localidad  \
104  CIUDAD AUTÓNOMA DE BUENOS AIRES                2                -34.6037   
0                       BUENOS AIRES                6                -34.7565   
67                      BUENOS AIRES                6                -38.1367   
29                   NO HUBO REQUEST  NO HUBO REQUEST         NO HUBO REQUEST   
40                      BUENOS AIRES                6                -34.8609   

    centroide_lon_localidad centroide_lat_municipio centroide_lon_municipio  \
104                -58.3805                    None                    None   
0                  -58.4903                    None                    None   
67                 -61.7957                    None                    None   
29          NO HUBO REQUEST                    None                    None   
40                 -58.3891                    None                    None   

    centroide_lat_departamento centroide_lon_departamento  \
104                       None                       None   
0                         None                       None   
67                        None                       None   
29                        None                       None   
40                        None                       None   

    centroide_lat_provincia centroide_lon_provincia  
104                    None                    None  
0                      None                    None  
67                     None                    None  
29                     None                    None  
40                     None                    None

In [30]:
gby_prov_loc['copy_index'] = gby_prov_loc.index

#Asignar centroide de municipios
gby_prov_loc.municipio_id_API = gby_prov_loc.municipio_id_API.apply(lambda x: (int(x) if  (x != 'NO HUBO REQUEST') and pd.notna(x) else x)) # Pasar a integer la columna
gby_prov_loc.centroide_lat_municipio = gby_prov_loc.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lat']
gby_prov_loc.centroide_lon_municipio = gby_prov_loc.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lon']

#Asignar centroide de departamentos
gby_prov_loc.departamento_id_API = gby_prov_loc.departamento_id_API.apply(lambda x: (int(x) if  x != 'NO HUBO REQUEST' else x)) # Pasar a integer la columna
gby_prov_loc.centroide_lat_departamento = gby_prov_loc.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lat']
gby_prov_loc.centroide_lon_departamento = gby_prov_loc.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lon']

#Asignar centroide de provincias
gby_prov_loc.provincia_id_API = gby_prov_loc.provincia_id_API.apply(lambda x: (int(x) if  x != 'NO HUBO REQUEST' else x)) # Pasar a integer la columna
gby_prov_loc.centroide_lat_provincia = gby_prov_loc.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lat']
gby_prov_loc.centroide_lon_provincia = gby_prov_loc.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_prov_loc['copy_index'])['centroide_lon']

In [31]:
gby_prov_loc.sample(5)

provincia_por_cp       localidad_por_cp  \
10                     BUENOS AIRES                CASEROS   
31                     BUENOS AIRES      JOSE CLEMENTE PAZ   
60                     BUENOS AIRES  SAN BERNARDO DEL TUYU   
82                     BUENOS AIRES     VILLA SANTOS TESEI   
95  CIUDAD AUTONOMA DE BUENOS AIRES                LA BOCA   

     localidad_por_API localidad_id_API municipio_por_API municipio_id_API  \
10             CASEROS      06840010001   TRES DE FEBRERO            60840   
31     NO HUBO REQUEST  NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
60     NO HUBO REQUEST  NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
82  VILLA SANTOS TESEI      06408010002        HURLINGHAM            60408   
95                BOCA      02028010002   NO HUBO REQUEST             None   

   departamento_por_API departamento_id_API                provincia_por_API  \
10      TRES DE FEBRERO                6840                     BUENOS AIRES   
31      NO HUBO REQUEST     NO HUBO REQUEST                  NO HUBO REQUEST   
60      NO HUBO REQUEST     NO HUBO REQUEST                  NO HUBO REQUEST   
82           HURLINGHAM                6408                     BUENOS AIRES   
95             COMUNA 4                2028  CIUDAD AUTÓNOMA DE BUENOS AIRES   

   provincia_id_API centroide_lat_localidad centroide_lon_localidad  \
10                6                -34.6089                -58.5637   
31  NO HUBO REQUEST         NO HUBO REQUEST         NO HUBO REQUEST   
60  NO HUBO REQUEST         NO HUBO REQUEST         NO HUBO REQUEST   
82                6                -34.6124                -58.6538   
95                2                -34.6311                -58.3568   

    centroide_lat_municipio  centroide_lon_municipio  \
10               -34.595980               -58.579093   
31                      NaN                      NaN   
60                      NaN                      NaN   
82               -34.599214               -58.649694   
95                      NaN                      NaN   

    centroide_lat_departamento  centroide_lon_departamento  \
10                  -34.595980                  -58.579093   
31                         NaN                         NaN   
60                         NaN                         NaN   
82                  -34.599214                  -58.649694   
95                  -34.642079                  -58.387455   

    centroide_lat_provincia  centroide_lon_provincia  copy_index  
10               -36.676942               -60.558832          10  
31                      NaN                      NaN          31  
60                      NaN                      NaN          60  
82               -36.676942               -60.558832          82  
95               -34.614493               -58.445856          95

#### Agregar los datos a la tabla aumentada

In [32]:
pd.set_option('display.max_columns', 50)  # Para ver todas las filas
# Vamos a hacer un Join con los valores que tengan la misma provincia y localidad por CP
augmented_df['copy_index'] = augmented_df.index.copy()  # Guardar los index iniciales
gby_prov_loc.drop(columns= 'copy_index', inplace= True)

augmented_df = augmented_df.merge(gby_prov_loc , on = ['provincia_por_cp','localidad_por_cp'],
                                  how= 'left', suffixes= ['_del','']).set_index(augmented_df['copy_index']).copy()
                                                                                 
augmented_df = augmented_df.drop(columns = augmented_df.filter(regex = '._del').columns)
augmented_df.drop(columns = ['copy_index'], inplace = True)

In [33]:
augmented_df.sample(5)

anio_ing nacionalidad provincia_part    loc_part cp_part  \
copy_index                                                             
302             2014           AR             BA     FLORIDA    1602   
69058           2019           AR             BA   ITUZAINGO    1714   
34278           2016           AR             CF        CABA    1424   
61012           2018           AR             BA  SAN ISIDRO    1642   
71034           2018           AR             BA  LA TABLADA    1766   

             user_id                 provincia_por_cp    localidad_por_cp  \
copy_index                                                                  
302           ID_302                     BUENOS AIRES  NO UNICA LOCALIDAD   
69058       ID_69058                     BUENOS AIRES           ITUZAINGO   
34278       ID_34278  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
61012       ID_61012                     BUENOS AIRES          SAN ISIDRO   
71034       ID_71034                     BUENOS AIRES             TABLADA   

              departamento_por_cp localidad_por_API localidad_id_API  \
copy_index                                                             
302                          None               NaN              NaN   
69058                        None  ITUZAINGO CENTRO      06410010001   
34278       NO UNICO DEPARTAMENTO               NaN              NaN   
61012                        None        SAN ISIDRO      06756010005   
71034                        None        LA TABLADA      06427010006   

           municipio_por_API municipio_id_API departamento_por_API  \
copy_index                                                           
302                      NaN              NaN                  NaN   
69058              ITUZAINGÓ            60410            ITUZAINGÓ   
34278                    NaN              NaN                  NaN   
61012             SAN ISIDRO            60756           SAN ISIDRO   
71034             LA MATANZA            60427           LA MATANZA   

           departamento_id_API provincia_por_API provincia_id_API  \
copy_index                                                          
302                        NaN               NaN              NaN   
69058                     6410      BUENOS AIRES                6   
34278                      NaN               NaN              NaN   
61012                     6756      BUENOS AIRES                6   
71034                     6427      BUENOS AIRES                6   

           centroide_lat_localidad centroide_lon_localidad  \
copy_index                                                   
302                            NaN                     NaN   
69058                     -34.6441                -58.6885   
34278                          NaN                     NaN   
61012                      -34.477                -58.5272   
71034                     -34.6871                -58.5256   

            centroide_lat_municipio  centroide_lon_municipio  \
copy_index                                                     
302                             NaN                      NaN   
69058                    -34.636002               -58.688758   
34278                           NaN                      NaN   
61012                    -34.486950               -58.537265   
71034                    -34.770165               -58.625449   

            centroide_lat_departamento  centroide_lon_departamento  \
copy_index                                                           
302                                NaN                         NaN   
69058                       -34.636002                  -58.688758   
34278                              NaN                         NaN   
61012                       -34.486950                  -58.537265   
71034                       -34.770165                  -58.625449   

            centroide_lat_provincia  centroide_lon_provincia  
copy_index                                       

### Conseguir datos de los que tienen solo provincia por codigo postal

In [34]:
gby_solo_provincia = augmented_df.groupby(['loc_part', 'provincia_por_cp','localidad_por_cp']).count(
                        ).reset_index(['loc_part','provincia_por_cp','localidad_por_cp']) # con este subset armar un post request
# Selecciono las opciones en las cuales el CP no dio una localidad definida pero si me dio provincia
gby_solo_provincia = gby_solo_provincia.loc[(gby_solo_provincia.provincia_por_cp != 'CP INDEFINIDO') &
                                            (gby_solo_provincia.provincia_por_cp != 'NO UNICA PROVINCIA') &
                                            (gby_solo_provincia.localidad_por_cp == 'NO UNICA LOCALIDAD')]
print(gby_solo_provincia.shape)
gby_solo_provincia.head(5)

(534, 25)


loc_part                 provincia_por_cp    localidad_por_cp  \
0    VENEZUELA Y PERU                     BUENOS AIRES  NO UNICA LOCALIDAD   
7                  -1                     BUENOS AIRES  NO UNICA LOCALIDAD   
13                 -1  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
15                 -1                         SAN JUAN  NO UNICA LOCALIDAD   
16                 -1                         SAN LUIS  NO UNICA LOCALIDAD   

    anio_ing  nacionalidad  provincia_part  cp_part  user_id  \
0          1             1               1        1        1   
7         22            22              22       22       22   
13         2             2               2        2        2   
15         1             1               1        1        1   
16         2             2               2        2        2   

    departamento_por_cp  localidad_por_API  localidad_id_API  \
0                     0                  0                 0   
7                     0                  0                 0   
13                    2                  0                 0   
15                    0                  0                 0   
16                    0                  0                 0   

    municipio_por_API  municipio_id_API  departamento_por_API  \
0                   0                 0                     0   
7                   0                 0                     0   
13                  0                 0                     0   
15                  0                 0                     0   
16                  0                 0                     0   

    departamento_id_API  provincia_por_API  provincia_id_API  \
0                     0                  0                 0   
7                     0                  0                 0   
13                    0                  0                 0   
15                    0                  0                 0   
16                    0                  0                 0   

    centroide_lat_localidad  centroide_lon_localidad  centroide_lat_municipio  \
0                         0                        0                        0   
7                         0                        0                        0   
13                        0                        0                        0   
15                        0                        0                        0   
16                        0                        0                        0   

    centroide_lon_municipio  centroide_lat_departamento  \
0                         0                           0   
7                         0                           0   
13                        0                           0   
15                        0                           0   
16                        0                           0   

    centroide_lon_departamento  centroide_lat_provincia  \
0                            0                        0   
7                            0                        0   
13                           0                        0   
15                           0                        0   
16                           0                        0   

    centroide_lon_provincia  
0                         0  
7                         0  
13                        0  
15                        0  
16                        0

In [35]:
# Crea una lista de tuplas para realizar el post
prov_loc = list(gby_solo_provincia.apply(lambda x: (x.provincia_por_cp, x.loc_part), axis=1))  
#En vez de localidad usamos loc_part (localidad reportada)
# Manda el POST request
datos = get_centroides.post_centroides('localidades', prov_loc, prov = True) # Consulta get_centroides.py

In [36]:
len(datos)

534

In [37]:
# Rellena la tabla con los datos
gby_solo_provincia = rellenar_df(gby_solo_provincia,datos)
gby_solo_provincia.sample(5)

loc_part                 provincia_por_cp    localidad_por_cp  \
204           CAPITAL                     BUENOS AIRES  NO UNICA LOCALIDAD   
546  GENERAL BELGRANO  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
824   LOMAS DE ZAMORA                     BUENOS AIRES  NO UNICA LOCALIDAD   
13                 -1  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
739            LA PAZ                       ENTRE RÍOS  NO UNICA LOCALIDAD   

     anio_ing  nacionalidad  provincia_part  cp_part  user_id  \
204         1             1               1        1        1   
546         1             1               1        1        1   
824        37            37              36       37       37   
13          2             2               2        2        2   
739         1             1               1        1        1   

     departamento_por_cp  localidad_por_API localidad_id_API  \
204                    0  CAPITAN SARMIENTO      06140010000   
546                    1    NO HUBO REQUEST  NO HUBO REQUEST   
824                    0    LOMAS DE ZAMORA      06490010003   
13                     2    NO HUBO REQUEST  NO HUBO REQUEST   
739                    0             LA PAZ      30070040000   

     municipio_por_API municipio_id_API departamento_por_API  \
204  CAPITÁN SARMIENTO            60140    CAPITÁN SARMIENTO   
546    NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
824    LOMAS DE ZAMORA            60490      LOMAS DE ZAMORA   
13     NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
739    NO HUBO REQUEST             None               LA PAZ   

    departamento_id_API provincia_por_API provincia_id_API  \
204                6140      BUENOS AIRES                6   
546     NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
824                6490      BUENOS AIRES                6   
13      NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
739               30070        ENTRE RÍOS               30   

    centroide_lat_localidad centroide_lon_localidad  centroide_lat_municipio  \
204                -34.1724                -59.7893                        0   
546         NO HUBO REQUEST         NO HUBO REQUEST                        0   
824                -34.7628                -58.4268                        0   
13          NO HUBO REQUEST         NO HUBO REQUEST                        0   
739                -30.7415                -59.6434                        0   

     centroide_lon_municipio  centroide_lat_departamento  \
204                        0                           0   
546                        0                           0   
824                        0                           0   
13                         0                           0   
739                        0                           0   

     centroide_lon_departamento  centroide_lat_provincia  \
204                           0                        0   
546                           0                        0   
824                           0                        0   
13                            0                        0   
739                           0                        0   

     centroide_lon_provincia  
204                        0  
546                        0  
824                        0  
13                         0  
739                        0

In [38]:
gby_solo_provincia['copy_index_2'] = gby_solo_provincia.index

# Asignar centroide de municipios
gby_solo_provincia.centroide_lat_municipio = gby_solo_provincia.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_solo_provincia['copy_index_2'])['centroide_lat']
gby_solo_provincia.centroide_lon_municipio = gby_solo_provincia.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_solo_provincia['copy_index_2'])['centroide_lon']
# Asignar centroide de departamentos
gby_solo_provincia.centroide_lon_departamento = gby_solo_provincia.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how ='left').set_index(gby_solo_provincia['copy_index_2'])['centroide_lon']
gby_solo_provincia.centroide_lat_departamento = gby_solo_provincia.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how='left').set_index(gby_solo_provincia['copy_index_2'])['centroide_lat']
# Asignar centroide de provincias
gby_solo_provincia.centroide_lat_provincia = gby_solo_provincia.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_solo_provincia['copy_index_2'])['centroide_lat']
gby_solo_provincia.centroide_lon_provincia = gby_solo_provincia.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_solo_provincia['copy_index_2'])['centroide_lon']

In [39]:
gby_solo_provincia.sample(5) 

loc_part provincia_por_cp    localidad_por_cp  anio_ing  \
445              DOCK SUD     BUENOS AIRES  NO UNICA LOCALIDAD         3   
1486             VIRREYES     BUENOS AIRES  NO UNICA LOCALIDAD        38   
739                LA PAZ       ENTRE RÍOS  NO UNICA LOCALIDAD         1   
1410  VILLA GENERAL BELGR          CÓRDOBA  NO UNICA LOCALIDAD         1   
44     ADROGUE-ALTE BROWN     BUENOS AIRES  NO UNICA LOCALIDAD         1   

      nacionalidad  provincia_part  cp_part  user_id  departamento_por_cp  \
445              3               2        3        3                    0   
1486            38              38       38       38                    0   
739              1               1        1        1                    0   
1410             1               1        1        1                    0   
44               1               1        1        1                    0   

           localidad_por_API localidad_id_API       municipio_por_API  \
445                 DOCK SUD      06035010004              AVELLANEDA   
1486                VIRREYES      06749010003            SAN FERNANDO   
739                   LA PAZ      30070040000         NO HUBO REQUEST   
1410  VILLA GENERAL BELGRANO      14007310000  VILLA GENERAL BELGRANO   
44           NO HUBO REQUEST  NO HUBO REQUEST         NO HUBO REQUEST   

     municipio_id_API departamento_por_API departamento_id_API  \
445             60035           AVELLANEDA                6035   
1486            60749         SAN FERNANDO                6749   
739              None               LA PAZ               30070   
1410           140049          CALAMUCHITA               14007   
44    NO HUBO REQUEST      NO HUBO REQUEST     NO HUBO REQUEST   

     provincia_por_API provincia_id_API centroide_lat_localidad  \
445       BUENOS AIRES                6                -34.6491   
1486      BUENOS AIRES                6                -34.4636   
739         ENTRE RÍOS               30                -30.7415   
1410           CÓRDOBA               14                -31.9808   
44     NO HUBO REQUEST  NO HUBO REQUEST         NO HUBO REQUEST   

     centroide_lon_localidad  centroide_lat_municipio  \
445                 -58.3412               -34.678206   
1486                -58.5762               -34.151270   
739                 -59.6434                      NaN   
1410                -64.5606               -31.984374   
44           NO HUBO REQUEST                      NaN   

      centroide_lon_municipio  centroide_lat_departamento  \
445                -58.341117                  -34.678206   
1486               -58.545954                  -34.151270   
739                       NaN                  -30.897456   
1410               -64.585415                  -32.202712   
44                        NaN                         NaN   

      centroide_lon_departamento  centroide_lat_provincia  \
445                   -58.341117               -36.676942   
1486                  -58.545954               -36.676942   
739                   -59.500449               -32.058874   
1410                  -64.616717               -32.142933   
44                           NaN                      NaN   

      centroide_lon_provincia  copy_index_2  
445                -60.558832           445  
1486               -60.558832          1486  
739                -59.201448           739  
1410               -63.801753          1410  
44                        NaN            44

###  Conseguir datos de los que no obtuvimos ni provincia ni localidad


In [40]:
gby_solo_loc_part = augmented_df.groupby(['loc_part', 'provincia_por_cp','localidad_por_cp']).count(
                        ).reset_index(['loc_part','provincia_por_cp','localidad_por_cp']) # con este subset armar un post request
# Selecciono las opciones en las cuales el CP no dio una localidad definida, ni provincia o tambien si el CP es indefinido
gby_solo_loc_part = gby_solo_loc_part.loc[(gby_solo_loc_part.provincia_por_cp == 'CP INDEFINIDO') |
                                            (gby_solo_loc_part.provincia_por_cp == 'NO UNICA PROVINCIA') &
                                            (gby_solo_loc_part.localidad_por_cp == 'NO UNICA LOCALIDAD')]
print(gby_solo_loc_part.shape)
gby_solo_loc_part.sample(5) #221 rows unicas

(221, 25)


loc_part provincia_por_cp localidad_por_cp  anio_ing  \
198             CANNING    CP INDEFINIDO    CP INDEFINIDO        15   
952               MUNRO    CP INDEFINIDO    CP INDEFINIDO         1   
844      LOS POLVORINES    CP INDEFINIDO    CP INDEFINIDO         1   
1106  RINCON DE MILBERG    CP INDEFINIDO    CP INDEFINIDO         1   
1109         RIO GRANDE    CP INDEFINIDO    CP INDEFINIDO         1   

      nacionalidad  provincia_part  cp_part  user_id  departamento_por_cp  \
198             15              15       15       15                    0   
952              1               1        1        1                    0   
844              1               0        1        1                    0   
1106             1               1        1        1                    0   
1109             1               1        1        1                    0   

      localidad_por_API  localidad_id_API  municipio_por_API  \
198                   0                 0                  0   
952                   0                 0                  0   
844                   0                 0                  0   
1106                  0                 0                  0   
1109                  0                 0                  0   

      municipio_id_API  departamento_por_API  departamento_id_API  \
198                  0                     0                    0   
952                  0                     0                    0   
844                  0                     0                    0   
1106                 0                     0                    0   
1109                 0                     0                    0   

      provincia_por_API  provincia_id_API  centroide_lat_localidad  \
198                   0                 0                        0   
952                   0                 0                        0   
844                   0                 0                        0   
1106                  0                 0                        0   
1109                  0                 0                        0   

      centroide_lon_localidad  centroide_lat_municipio  \
198                         0                        0   
952                         0                        0   
844                         0                        0   
1106                        0                        0   
1109                        0                        0   

      centroide_lon_municipio  centroide_lat_departamento  \
198                         0                           0   
952                         0                           0   
844                         0                           0   
1106                        0                           0   
1109                        0                           0   

      centroide_lon_departamento  centroide_lat_provincia  \
198                            0                        0   
952                            0                        0   
844                            0                        0   
1106                           0                        0   
1109                           0                        0   

      centroide_lon_provincia  
198                         0  
952                         0  
844                         0  
1106                        0  
1109                        0

In [41]:
# Crea una lista de tuplas para realizar el post
prov_loc = list(gby_solo_loc_part.apply(lambda x: (x.provincia_por_cp, x.loc_part), axis=1))  # no necesito saber que no se sabe la localidad
# En vez de localidad usamos loc_part (localidad reportada)
# Manda el POST request
datos = get_centroides.post_centroides('localidades', prov_loc, prov = False) # Consulta get_centroides.py, no uso la provincia en este caso

In [42]:
len(datos)

221

In [43]:
# Rellena la tabla con los datos
gby_solo_loc_part = rellenar_df(gby_solo_loc_part,datos)
gby_solo_loc_part.sample(5)

loc_part provincia_por_cp localidad_por_cp  anio_ing  \
1127         SAENZ PEÑA    CP INDEFINIDO    CP INDEFINIDO         5   
687           ITUZAINGO    CP INDEFINIDO    CP INDEFINIDO        37   
1021              PILAR    CP INDEFINIDO    CP INDEFINIDO         3   
14                   -1    CP INDEFINIDO    CP INDEFINIDO        24   
1317  TRONCOS DEL TALAR    CP INDEFINIDO    CP INDEFINIDO         8   

      nacionalidad  provincia_part  cp_part  user_id  departamento_por_cp  \
1127             5               5        5        5                    0   
687             37              27       37       37                    0   
1021             3               3        3        3                    0   
14              24              24       24       24                    0   
1317             8               8        8        8                    0   

          localidad_por_API localidad_id_API       municipio_por_API  \
1127             SAENZ PEÑA      06840010012         TRES DE FEBRERO   
687               ITUZAINGO      18084020000               ITUZAINGÓ   
1021                  PILAR      06638040008                   PILAR   
14              1 DE AGOSTO      14021310004  ESTACIÓN JUÁREZ CELMAN   
1317  LOS TRONCOS DEL TALAR      06805010007                   TIGRE   

     municipio_id_API departamento_por_API departamento_id_API  \
1127            60840      TRES DE FEBRERO                6840   
687            180168            ITUZAINGÓ               18084   
1021            60638                PILAR                6638   
14             140105                COLÓN               14021   
1317            60805                TIGRE                6805   

     provincia_por_API provincia_id_API centroide_lat_localidad  \
1127      BUENOS AIRES                6                -34.5993   
687         CORRIENTES               18                 -27.591   
1021      BUENOS AIRES                6                -34.4663   
14             CÓRDOBA               14                -31.3069   
1317      BUENOS AIRES                6                -34.4426   

     centroide_lon_localidad  centroide_lat_municipio  \
1127                -58.5321                        0   
687                  -56.704                        0   
1021                -58.9227                        0   
14                  -64.1785                        0   
1317                -58.6174                        0   

      centroide_lon_municipio  centroide_lat_departamento  \
1127                        0                           0   
687                         0                           0   
1021                        0                           0   
14                          0                           0   
1317                        0                           0   

      centroide_lon_departamento  centroide_lat_provincia  \
1127                           0                        0   
687                            0                        0   
1021                           0                        0   
14                             0                        0   
1317                           0                        0   

      centroide_lon_provincia  
1127                        0  
687                         0  
1021                        0  
14                          0  
1317                        0

In [44]:
gby_solo_loc_part['copy_index_2'] = gby_solo_loc_part.index
# Asignar centroide de municipios
gby_solo_loc_part.centroide_lat_municipio = gby_solo_loc_part.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_solo_loc_part['copy_index_2'])['centroide_lat']
gby_solo_loc_part.centroide_lon_municipio = gby_solo_loc_part.merge(municipios, left_on= 'municipio_id_API', right_on='id', how='left' ).set_index(gby_solo_loc_part['copy_index_2'])['centroide_lon']
# Asignar centroide de departamentos
gby_solo_loc_part.centroide_lon_departamento = gby_solo_loc_part.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how ='left').set_index(gby_solo_loc_part['copy_index_2'])['centroide_lon']
gby_solo_loc_part.centroide_lat_departamento = gby_solo_loc_part.merge(departamentos, left_on= 'departamento_id_API', right_on='id', how='left').set_index(gby_solo_loc_part['copy_index_2'])['centroide_lat']
# Asignar centroide de provincias
gby_solo_loc_part.centroide_lat_provincia = gby_solo_loc_part.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_solo_loc_part['copy_index_2'])['centroide_lat']
gby_solo_loc_part.centroide_lon_provincia = gby_solo_loc_part.merge(provincias, left_on= 'provincia_id_API', right_on='id', how='left' ).set_index(gby_solo_loc_part['copy_index_2'])['centroide_lon']
gby_solo_loc_part.sample(10)

loc_part    provincia_por_cp    localidad_por_cp  anio_ing  \
1388          VILLA CAROLA       CP INDEFINIDO       CP INDEFINIDO         3   
1127            SAENZ PEÑA       CP INDEFINIDO       CP INDEFINIDO         5   
1266               TABLADA       CP INDEFINIDO       CP INDEFINIDO        21   
687              ITUZAINGO       CP INDEFINIDO       CP INDEFINIDO        37   
917         MONTE CHINGOLO       CP INDEFINIDO       CP INDEFINIDO         4   
1038               POSADAS  NO UNICA PROVINCIA  NO UNICA LOCALIDAD        14   
1104         RICARDO ROJAS       CP INDEFINIDO       CP INDEFINIDO         2   
91              AVELLANEDA       CP INDEFINIDO       CP INDEFINIDO        49   
845   LOTE 86 NEUQUEN CAP.  NO UNICA PROVINCIA  NO UNICA LOCALIDAD         1   
1496                 WILDE       CP INDEFINIDO       CP INDEFINIDO         3   

      nacionalidad  provincia_part  cp_part  user_id  departamento_por_cp  \
1388             3               3        3        3                    0   
1127             5               5        5        5                    0   
1266            21              14       21       21                    0   
687             37              27       37       37                    0   
917              4               4        4        4                    0   
1038            14              14       14       14                    0   
1104             2               2        2        2                    0   
91              49              16       49       49                    0   
845              1               1        1        1                    0   
1496             3               1        3        3                    0   

     localidad_por_API localidad_id_API municipio_por_API municipio_id_API  \
1388   NO HUBO REQUEST  NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
1127        SAENZ PEÑA      06840010012   TRES DE FEBRERO            60840   
1266        LA TABLADA      06427010006        LA MATANZA            60427   
687          ITUZAINGO      18084020000         ITUZAINGÓ           180168   
917     MONTE CHINGOLO      06434010004             LANÚS            60434   
1038           POSADAS      54028030000           POSADAS           540119   
1104     RICARDO ROJAS      06805010008             TIGRE            60805   
91          AVELLANEDA      06035010002        AVELLANEDA            60035   
845    NO HUBO REQUEST  NO HUBO REQUEST   NO HUBO REQUEST  NO HUBO REQUEST   
1496             WILDE      06035010009        AVELLANEDA            60035   

     departamento_por_API departamento_id_API provincia_por_API  \
1388      NO HUBO REQUEST     NO HUBO REQUEST   NO HUBO REQUEST   
1127      TRES DE FEBRERO                6840      BUENOS AIRES   
1266           LA MATANZA                6427      BUENOS AIRES   
687             ITUZAINGÓ               18084        CORRIENTES   
917                 LANÚS                6434      BUENOS AIRES   
1038              CAPITAL               54028          MISIONES   
1104                TIGRE                6805      BUENOS AIRES   
91             AVELLANEDA                6035      BUENOS AIRES   
845       NO HUBO REQUEST     NO HUBO REQUEST   NO HUBO REQUEST   
1496           AVELLANEDA                6035      BUENOS AIRES   

     provincia_id_API centroide_lat_localidad centroide_lon_localidad  \
1388  NO HUBO REQUEST         NO HUBO REQUEST         NO HUBO REQUEST   
1127                6                -34.5993                -58.5321   
1266                6                -34.6871                -58.5256   
687                18                 -27.591                 -56.704   
917                 6                -34.7302                -58.3562   
1038               54                -27.3662                 -55.894   
1104                6                -34.4563                -58.6807   
91                  6                -34.6627                -58.3657   
845   NO HUBO REQUEST         NO HU

In [45]:
gby_solo_loc_part[gby_solo_loc_part['localidad_por_API'] == 'NO HUBO REQUEST'].shape #34 request no devolvieron resultados

(34, 26)

## Agregar todos los datos a la tabla

In [46]:
aux_df = gby_solo_loc_part.append(gby_solo_provincia, sort = False) #append all tables in one df. Except prov_loc 

In [47]:
# Vamos a hacer un mege con los valores que tengan la misma provincia y localidad por CP y loc_part
augmented_df['copy_index'] = augmented_df.index.copy()  # Guardar los index iniciales

augmented_df2 = augmented_df.merge(aux_df ,
                                   on = ['loc_part','provincia_por_cp','localidad_por_cp'],
                                   how= 'left',
                                   suffixes= ['','_added']).set_index(augmented_df['copy_index']).copy()


In [48]:
def assign_value(row, column = None):
    '''
    Simple function that takes a row and a column and concatenate two columns after permorming a merge
    In this case the right table with the same column name was added a `_added` termination
    '''
    if pd.isna(row[column]):
        return row[column + '_added']
    else:
        return row[column]

In [49]:
col = ['provincia_por_API','localidad_por_API','departamento_por_API','municipio_por_API',
        'centroide_lat_localidad','centroide_lon_localidad','centroide_lat_departamento',
        'centroide_lon_departamento','centroide_lat_municipio','centroide_lon_municipio','centroide_lat_provincia',
        'centroide_lon_provincia','provincia_id_API','departamento_id_API','municipio_id_API','localidad_id_API']
for c in col:
    augmented_df2[c] = augmented_df2.apply(assign_value, column = c, axis = 1)

In [50]:
augmented_df2.drop(columns = augmented_df2.filter(regex = '._added').columns, inplace=True) #drop _added columns
augmented_df2.drop(columns = ['copy_index', 'copy_index_2'], inplace = True) #drop las columnas con indices

In [51]:
augmented_df2.sample(5)

anio_ing nacionalidad provincia_part             loc_part cp_part  \
copy_index                                                                      
61624           2018           UY             CF                 CABA    1225   
19356           2015           AR             BA    LOMAS DEL MIRADOR    1752   
59905           2017  DESCONOCIDO            NaN  CDAD AUTONOMA BS AS       0   
71692           2018           AR             CF                 CABA    1424   
34566           2016           AR             CF                 CABA    1407   

             user_id                 provincia_por_cp    localidad_por_cp  \
copy_index                                                                  
61624       ID_61624  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
19356       ID_19356                     BUENOS AIRES  NO UNICA LOCALIDAD   
59905       ID_59905                    CP INDEFINIDO       CP INDEFINIDO   
71692       ID_71692  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
34566       ID_34566  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   

              departamento_por_cp  localidad_por_API localidad_id_API  \
copy_index                                                              
61624       NO UNICO DEPARTAMENTO          CABALLITO      02042010001   
19356                        None  LOMAS DEL MIRADOR      06427010007   
59905                        None    NO HUBO REQUEST  NO HUBO REQUEST   
71692       NO UNICO DEPARTAMENTO          CABALLITO      02042010001   
34566       NO UNICO DEPARTAMENTO          CABALLITO      02042010001   

           municipio_por_API municipio_id_API departamento_por_API  \
copy_index                                                           
61624        NO HUBO REQUEST             None             COMUNA 6   
19356             LA MATANZA            60427           LA MATANZA   
59905        NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
71692        NO HUBO REQUEST             None             COMUNA 6   
34566        NO HUBO REQUEST             None             COMUNA 6   

           departamento_id_API                provincia_por_API  \
copy_index                                                        
61624                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
19356                     6427                     BUENOS AIRES   
59905          NO HUBO REQUEST                  NO HUBO REQUEST   
71692                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
34566                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   

           provincia_id_API centroide_lat_localidad centroide_lon_localidad  \
copy_index                                                                    
61624                     2                -34.6168                -58.4436   
19356                     6                -34.6657                -58.5332   
59905       NO HUBO REQUEST         NO HUBO REQUEST         NO HUBO REQUEST   
71692                     2                -34.6168                -58.4436   
34566                     2                -34.6168                -58.4436   

            centroide_lat_municipio  centroide_lon_municipio  \
copy_index                                                     
61624                           NaN                      NaN   
19356                    -34.770165               -58.625449   
59905                           NaN                      NaN   
71692                           NaN                      NaN   
34566                           NaN                      NaN   

            centroide_lat_departamento  centroide_lon_departamento  \
copy_index                                                           
61624                       -34.616843                  -58.443568   
19356                       -34.770165                  -58.625449   
59905                              NaN                         NaN   
71692                       -34.616843                  -58.443568   
34566   

## EDA

In [52]:
augmented_df2[augmented_df2['localidad_por_API'] == 'NO HUBO REQUEST'].shape

(6877, 25)

In [53]:
augmented_df2.shape

(87402, 25)

In [54]:
augmented_df2[augmented_df2['localidad_por_API'] == 'NO HUBO REQUEST'].shape[0]/augmented_df2.shape[0] * 100

7.868241001350083

## Analisis de los datos

Aproximadamente un 7.8% no pudo ser identificado.<br>
Tenemos que mencionar que confiar en el codigo postal lejos esta de ser el mejor de los casos (los sujetos pueden olvidarlo o ingresarlo erroneamente sin darse cuenta) pero frente los inconvenientes debidos a la falta de normalizacion de la base inicial nos parecio la mejor opcion. 
Hay que aclarar que otro problema es que los usuarios de CABA usualmente no ponen la localidad en el campo `loc_part` por lo que deberiamos hacer un sistema doble en el cual se asigne el campo de `provincia_por_cp` a `provincia_por_API`

---

# Modificacion de anomalias

In [55]:
augmented_df2[augmented_df2.localidad_por_API == 'CABALLITO'].sample(10) 
#  Debo arreglar que todos los que dice CABA y no tengan localidad por CP pasen a ser solo Capital

anio_ing nacionalidad provincia_part loc_part cp_part   user_id  \
copy_index                                                                    
33258           2016           SE             CF     CABA    1419  ID_33258   
61111           2018           AR             CF     CABA    1424  ID_61111   
66775           2018           AR             CF     CABA    1431  ID_66775   
72372           2018           AR             CF     CABA    1405  ID_72372   
27316           2016           PY             CF     CABA    1426  ID_27316   
30674           2016           AR             CF     CABA    1417  ID_30674   
70882           2018           AR             CF     CABA    1428  ID_70882   
73437           2018           AR             CF     CABA    1184  ID_73437   
11719           2014           AR             CF     CABA    1417  ID_11719   
54126           2017           AR             CF     CABA    1184  ID_54126   

                           provincia_por_cp    localidad_por_cp  \
copy_index                                                        
33258       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
61111       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
66775       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
72372       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
27316       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
30674       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
70882       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
73437       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
11719       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
54126       CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   

              departamento_por_cp localidad_por_API localidad_id_API  \
copy_index                                                             
33258       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
61111       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
66775       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
72372       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
27316       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
30674       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
70882       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
73437       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
11719       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   
54126       NO UNICO DEPARTAMENTO         CABALLITO      02042010001   

           municipio_por_API municipio_id_API departamento_por_API  \
copy_index                                                           
33258        NO HUBO REQUEST             None             COMUNA 6   
61111        NO HUBO REQUEST             None             COMUNA 6   
66775        NO HUBO REQUEST             None             COMUNA 6   
72372        NO HUBO REQUEST             None             COMUNA 6   
27316        NO HUBO REQUEST             None             COMUNA 6   
30674        NO HUBO REQUEST             None             COMUNA 6   
70882        NO HUBO REQUEST             None             COMUNA 6   
73437        NO HUBO REQUEST             None             COMUNA 6   
11719        NO HUBO REQUEST             None             COMUNA 6   
54126        NO HUBO REQUEST             None             COMUNA 6   

           departamento_id_API                provincia_por_API  \
copy_index                                                        
33258                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
61111                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
66775                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
72372                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
27316                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
30674                     2042  CIUDAD AUTÓNOMA DE BUENOS AIRES   
70882           

In [56]:
mask = augmented_df2[(augmented_df2.loc_part == 'CABA')  &
              (augmented_df2.localidad_por_cp == 'NO UNICA LOCALIDAD') &
              (augmented_df2.provincia_por_cp == 'CIUDAD AUTONOMA DE BUENOS AIRES')].index

In [57]:
cols_modif = [9,10,11,12,13,14,17,18,19,20,21,22]
augmented_df2.iloc[mask,cols_modif] = 'NO HUBO REQUEST'

In [58]:
#  Debo arreglar que todos los que dicen `CDA AUTONOMA DE BS AS` o tengan localidad por CP pasen a ser solo Capital
mask = augmented_df2[(augmented_df2.loc_part == 'CDAD AUTONOMA BS AS') &
              (augmented_df2.localidad_por_cp == 'NO UNICA LOCALIDAD') &
              (augmented_df2.provincia_por_cp == 'CIUDAD AUTONOMA DE BUENOS AIRES')].index

augmented_df2.loc[mask,'localidad_id_API'] = 'NO HUBO REQUEST'
augmented_df2.loc[mask,'provincia_por_API'] = 'CIUDAD AUTÓNOMA DE BUENOS AIRES'
augmented_df2.loc[mask,'provincia_id_API'] = 2
augmented_df2.loc[mask,'centroide_lat_provincia'] = -34.614493
augmented_df2.loc[mask,'centroide_lon_provincia'] = -58.445856

In [59]:
augmented_df2

anio_ing nacionalidad provincia_part             loc_part cp_part  \
copy_index                                                                      
0               2014           AR             CF                 CABA    1416   
1               2014           AR             CF                 CABA    1416   
2               2014           AR             BA           HURLINGHAM    1686   
3               2014           AR             BA               ZARATE    2800   
4               2014           AR             CF                 CABA    1425   
...              ...          ...            ...                  ...     ...   
87397           2019           AR             CF  CDAD AUTONOMA BS AS    1417   
87398           2019           AR             CF  CDAD AUTONOMA BS AS    1182   
87399           2019           AR             CF  CDAD AUTONOMA BS AS    1440   
87400           2019           AR             CF  CDAD AUTONOMA BS AS    1424   
87401           2020           AR             CF  CDAD AUTONOMA BS AS    1425   

             user_id                 provincia_por_cp    localidad_por_cp  \
copy_index                                                                  
0               ID_0  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
1               ID_1  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
2               ID_2                     BUENOS AIRES          HURLINGHAM   
3               ID_3                     BUENOS AIRES  NO UNICA LOCALIDAD   
4               ID_4  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
...              ...                              ...                 ...   
87397       ID_87397  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
87398       ID_87398  CIUDAD AUTONOMA DE BUENOS AIRES             PALERMO   
87399       ID_87399  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
87400       ID_87400  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   
87401       ID_87401  CIUDAD AUTONOMA DE BUENOS AIRES  NO UNICA LOCALIDAD   

              departamento_por_cp localidad_por_API localidad_id_API  \
copy_index                                                             
0           NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
1           NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
2                            None        HURLINGHAM      06408010001   
3                            None            ZARATE      06882050000   
4           NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
...                           ...               ...              ...   
87397       NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
87398                   COMUNA 14           PALERMO      02098010001   
87399       NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
87400       NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   
87401       NO UNICO DEPARTAMENTO   NO HUBO REQUEST  NO HUBO REQUEST   

           municipio_por_API municipio_id_API departamento_por_API  \
copy_index                                                           
0            NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
1            NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
2                 HURLINGHAM            60408           HURLINGHAM   
3                     ZÁRATE            60882               ZÁRATE   
4            NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
...                      ...              ...                  ...   
87397        NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
87398        NO HUBO REQUEST              NaN            COMUNA 14   
87399        NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
87400        NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   
87401        NO HUBO REQUEST  NO HUBO REQUEST      NO HUBO REQUEST   

           departamento_id_API                provincia_por_API  \
copy_index                               

# Analisis de los datos post modificacion

In [64]:
print(augmented_df2[augmented_df2.provincia_por_API == 'NO HUBO REQUEST'].shape[0] / augmented_df2.shape[0] * 100)
print(augmented_df2[augmented_df2.localidad_por_API == 'NO HUBO REQUEST'].shape[0] / augmented_df2.shape[0] * 100)
print(augmented_df2[augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES'].shape[0] / augmented_df2.shape[0] * 100)

print(augmented_df2[(augmented_df2.localidad_por_API == 'NO HUBO REQUEST') &
                    (augmented_df2.provincia_por_API != 'CIUDAD AUTÓNOMA DE BUENOS AIRES')].shape[0] / augmented_df2.shape[0] * 100)

print(augmented_df2[(augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES') & (augmented_df2.localidad_por_API != 'NO HUBO REQUEST')].shape[0] /
      augmented_df2[augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES'].shape[0] * 100)

4.104025079517632
44.384567858859064
45.168302784833294
4.104025079517632
10.821216880287755


Posterior a las modificaciones solo el 4% del padron no tiene un campo para provincia. <br>
Un 44.3 % no tiene una localidad definida. De ese 44% la mayoria responde a personas de CABA (un 45.1 % del padron). Si excluimos a esta gente solo un 4.1 % del padron no tiene una localidad normalizada. <br>
Por otro lado, los usuarios de CABA que SI POSEEN localidad definida son solo el 10.8 % del padron de CABA


In [67]:
pickle.dump(augmented_df2, open('HI_localizacion_socios.pickle','wb'))

---

# Analisis de la implementacion de un algoritmo para geolocalizar usuarios

In [28]:
augmented_df2 = pickle.load(open('./HI_localizacion_socios.pickle', 'rb'))

## Comparacion por provincia

In [32]:
df_clean.isna().sum().provincia_part / df_clean.shape[0] * 100

5.512459669115123

In [45]:
# Total de Nan y `NO HUBO REQUEST` en tabla aumentada
total_miss = augmented_df2.provincia_por_API.isna().sum() + augmented_df2[augmented_df2.provincia_por_API == 'NO HUBO REQUEST'].count().provincia_por_API
total_miss / augmented_df2.shape[0] * 100

4.333996933708611

In [47]:
# Provincias nuevas incluidas
df_clean.isna().sum().provincia_part - total_miss

1030

Mientras que nuestro dataset inicial contaba con un 5.51% de valores missing para el campo de provincia, tras correr el algoritmo se alcanzo se paso a 4.33% de missing values. En otras palabras, tras correr el algoritmo se obtuvo un incremento de **1030** usuarios con un campo de provincia asignado.

## Comparacion por localidad

In [55]:
df_clean.isna().sum().loc_part / df_clean.shape[0] * 100

0.30090844603098327

In [56]:
(augmented_df2.localidad_por_API.isna().sum() +  augmented_df2[augmented_df2.localidad_por_API == 'NO HUBO REQUEST'].shape[0]) / df_clean.shape[0] * 100

44.61453971305004

A pesar de que solo el 0.3 % no presenta una localidad reportada, al intentar localizar la localidad usando nuestro algoritmo vemos que un 44% de los usuarios no tienen una localidad definida, normalizada y geolocalizada (con un centroide). 

In [67]:
print(f'Cantidad de usuarios de la provincia de Buenos Aires por API: {augmented_df2[(augmented_df2.provincia_por_API == "BUENOS AIRES")].shape[0]}')

Cantidad de usuarios de la provincia de Buenos Aires por API: 42447


In [77]:
print(f'Cantidad de usuarios de la provincia de Buenos Aires por API sin localidad asignada: {augmented_df2[(augmented_df2.localidad_por_API == "NO HUBO REQUEST") & (augmented_df2.provincia_por_API == "BUENOS AIRES")].shape[0]}')

Cantidad de usuarios de la provincia de Buenos Aires por API sin localidad asignada: 0


In [82]:
print(f"Cantidad de usuarios de CABA por API: {augmented_df2[augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES'].shape[0]}")

Cantidad de usuarios de CABA por API: 39478


In [65]:
print(f"Cantidad de usuarios de CABA por API sin localidad asignada: {augmented_df2[(augmented_df2.localidad_por_API == 'NO HUBO REQUEST') & (augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES')].shape[0]}")

Cantidad de usuarios de CABA por API sin localidad asignada: 35206


In [91]:
prom_miss_caba = augmented_df2[(augmented_df2.localidad_por_API == 'NO HUBO REQUEST') &
                               (augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES')].shape[0] * 100 \
                 / augmented_df2[(augmented_df2.provincia_por_API == 'CIUDAD AUTÓNOMA DE BUENOS AIRES')].shape[0] 

In [94]:
100 - prom_miss_caba

10.82121688028775

In [73]:
total_con_loc = augmented_df2[(augmented_df2.localidad_por_API.notna()) & (augmented_df2.localidad_por_API != 'NO HUBO REQUEST')].shape[0]
(augmented_df2.shape[0] - total_con_loc) / augmented_df2.shape[0] * 100

44.61453971305004

**Todos** los usuarios de la Provincia de **Buenos Aires (100%)** se les asigno una localidad normalizada y se los georeferencio (42447 usuarios).<br>
De los usuarios de la **Capital Federal** (39478 usuarios) solo se pudo asignar una localidad y geolocalizar al **10.8%** (4272 usuarios) <br>
En **total** un **44.6%** de los usuarios (48408) fueron **geolocalizados** y sus **localidades** fueron **normalizadas**.